### DISCUSSION TITLE
Time out while trying to send show version on a cisco router.

### DISCUSSION THREAD

**Discussion by sebad78** [scrapli/scrapligo#185] [Q&A] [answered]
The test function TestNewSSHService was created to obtain configuration information from routers. It was tested with two routers: a Cisco 1941 series and a Palo Alto PA-220 series.

```golang
func TestNewSSHService(t *testing.T) {

	extraKexs := []string{
		"diffie-hellman-group14-sha1",
		"diffie-hellman-group-exchange-sha256",
		"diffie-hellman-group14-sha",
	}
	extraCiphers := []string{"aes128-ctr", "aes192-ctr", "aes256-ctr", "aes128-cbc", "aes192-cbc", "aes256-cbc", "ssh-rsa"}
	driver, e := generic.NewDriver(
		"100.100.xxx.xxx",
		options.WithAuthNoStrictKey(),
		options.WithTransportType("standard"),
		options.WithStandardTransportExtraKexs(extraKexs),       // Their offer: diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha
		options.WithStandardTransportExtraCiphers(extraCiphers), // Their offer: aes128-cbc,aes192-cbc,aes256-cbc,aes128-ctr,aes192-ctr,aes256-ctr
		options.WithAuthUsername("user"),
		options.WithAuthPassword("password"),
		options.WithTimeoutSocket(15*time.Second),
		options.WithTimeoutOps(15*time.Second),
		options.WithDefaultLogger(),
	)
	if e != nil {
		t.Error(e)
	}

	if driver.Open() == nil {
		defer driver.Close()
	}

	r, err := driver.SendCommand("show version")
	if err != nil {
		t.Error(err)
		if err != nil {
			t.Fatalf("Error sending command to ios: %v, details: %v", r, err)
		}
	} else {
		t.Logf("result %s", r.Result)
	}

}
```

The test then attempts to establish an SSH connection with the router and send the "show version" command to retrieve its version information. However, while the connection to the Cisco router is established successfully, it times out after attempting to send the command.

<img width="985" alt="image" src="https://github.com/scrapli/scrapligo/assets/1242746/bad9541c-efc4-41b2-b6f4-55d23947415d">

In contrast, the test works fine with the Palo Alto router, where it successfully establishes the connection and retrieves the command result without any issues.

<img width="1000" alt="image" src="https://github.com/scrapli/scrapligo/assets/1242746/fbae9665-921b-4405-8abd-987082a64e4a">

Any help is greatly appreciated!

**Comment by carlmontanari** [comment]
👋 if I had to guess you dont have sufficient privileges with the user for the ios device but thats a guess. debug logging and/or (probably even better) channel log would make it very clear whats up. Check [this](https://github.com/scrapli/scrapligo/blob/main/examples/generic_driver/channel_logging/main.go) for channel log stuff.

**Reply by sebad78** [reply]
Thank you Carl for the quick response!!!!
l'm going to check the channel and let you know.

**Reply by sebad78** [reply]
This is the ouput, it looks like the channel gets the result:
```sh
got a log item of type string
2024/06/04 11:35:45    DEBUG opening connection to host '100.100.x.x' on port '22'
got a log item of type string
2024/06/04 11:35:48    DEBUG starting channel read loop
got a log item of type string
2024/06/04 11:35:48     INFO connection opened successfully
got a log item of type string
2024/06/04 11:35:48     INFO SendCommand requested, sending 'show version'
got a log item of type string
2024/06/04 11:35:48    DEBUG channel SendInput requested, sending input 'show version'
got a log item of type string
2024/06/04 11:35:48    DEBUG channel write "show version"
got a log item of type string
2024/06/04 11:35:48    DEBUG channel read "\nRouter_1941#"
got a log item of type string
2024/06/04 11:35:48    DEBUG channel read "sh"
got a log item of type string
2024/06/04 11:35:48    DEBUG channel read "ow version"
got a log item of type string
2024/06/04 11:35:48    DEBUG channel write "\n"
got a log item of type string
2024/06/04 11:35:48    DEBUG channel read "\n"
got a log item of type string
2024/06/04 11:35:48    DEBUG channel read "Cisco IOS Software, C1900 Software (C1900-UNIVERSALK9-M),
...
... 
ALL THE LOG 
Router_1941#"
got a log item of type string
2024/06/04 11:36:03 CRITICAL channel timeout sending input to device
    ssh_test.go:79: errTimeoutError: channel timeout sending input to device
    ssh_test.go:81: Error sending command to ios: <nil>, details: errTimeoutError: channel timeout sending input to device
got a log item of type string
2024/06/04 11:36:03    DEBUG closing connection to host '100.100.XX.XX' on port '22'
got a log item of type string
2024/06/04 11:36:03     INFO channel closing...
got a log item of type string
2024/06/04 11:36:03    DEBUG force closing underlying transport...
got a log item of type string
2024/06/04 11:36:03     INFO connection closed successfully
--- FAIL: TestNewSSHService (17.66s)
```

**Reply by carlmontanari** [reply, marked as answer]
ah, that underscore in the hostname is breaking stuff, [here](https://github.com/scrapli/scrapligo/blob/main/assets/platforms/cisco_iosxe.yaml#L16C17-L16C41) is the default pattern for iosxe and [here](https://github.com/scrapli/scrapligo/blob/main/channel/channel.go#L40C39-L40C75) for generic since it looks like you're just using that anyway. you can use [this](https://github.com/scrapli/scrapligo/blob/main/driver/options/channel.go#L33) to pass a custom pattern to the generic driver -- if you want to do the same for the platform specific iosxe driver then you can copy the default platform and tweak it as needed.

**Reply by sebad78** [reply]
thank you Carl!!!
